core/src/transport/mod.rs

```diff
@@ -47,73 +46,85 @@ pub use self::choice::OrTransport;
 pub use self::memory::MemoryTransport;
 pub use self::upgrade::Upgrade;
 
-/// A transport is an object that can be used to produce connections by listening or dialing a
-/// peer.
+/// A transport is a connection-oriented communication channel between two peers,
+/// providing ordered streams of data and (usually) reliable delivery.
+///
+/// This trait is implemented on concrete connection-oriented transport protocols
+/// like TCP or Unix Domain Sockets, but also on wrappers that add additional
+/// functionality to the dialing or listening process (e.g. name resolution via
+/// the DNS).
 ///
-/// This trait is implemented on concrete transports (e.g. TCP, UDP, etc.), but also on wrappers
-/// around them.
+/// Additional protocols can be layered on top of the connections established
+/// by a [`Transport`] through an upgrade mechanism that is initiated via
+/// [`with_upgrade`](Transport::with_upgrade) and optionally followed by further upgrades
+/// through chaining calls to [`with_upgrade`](Transport::with_upgrade) and
+/// [`and_then`](Transport::and_then). Thereby every upgrade yields a new [`Transport`]
+/// whose connection setup incorporates all earlier upgrades followed by the new upgrade,
+/// i.e. the order of the upgrades is significant.
 ///
 /// > **Note**: The methods of this trait use `self` and not `&self` or `&mut self`. In other
 /// >           words, listening or dialing consumes the transport object. This has been designed
 /// >           so that you would implement this trait on `&Foo` or `&mut Foo` instead of directly
 /// >           on `Foo`.
 pub trait Transport {
-    /// The raw connection to a peer.
+    /// A data stream (i.e. connection) created by the transport.
     type Output;
-    /// Error that can happen when dialing or listening.
+
+    /// An error that occurred during connection setup.
     type Error: error::Error;
 
-    /// The listener produces incoming connections.
+    /// A stream of incoming connections.
     ///
     /// An item should be produced whenever a connection is received at the lowest level of the
-    /// transport stack. The item is a `Future` that is signalled once some pre-processing has
-    /// taken place, and that connection has been upgraded to the wanted protocols.
+    /// transport stack. The item is a [`Future`] that resolves to a connection once all
+    /// protocol upgrades have been applied.
```

Comment:
The protocol upgrades are applied on the `ListenerUpgrade`. May need better wording.

Reply:
New suggestion in https://github.com/libp2p/rust-libp2p/commit/3216bae891144c5118fa90de9423ab3b878ceb6b.